File: Disjoint Set/Java/ImplementationRank.java
// Time: O(4 alpha)
public class DisjointSet {
    int[] rank;
    int[] ultimateParent;

    public DisjointSet(int num) {
        this.rank = new int[num + 1];
        this.ultimateParent = new int[num + 1];
        for(int index = 0; index <= num; index++) {
            this.ultimateParent[index] = index;
        }
    }

    private int findParent(int node) {
        if(node == this.ultimateParent[node]) {
            return node;
        }
        this.ultimateParent[node] = this.findParent(this.ultimateParent[node]);
        return this.ultimateParent[node];
    }

    private void unionByRank(int nodeOne, int nodeTwo) {
        int parentOne = this.findParent(nodeOne), parentTwo = this.findParent(nodeTwo);

        if(parentTwo == parentOne) {
            return;
        }

        int rankOne = this.rank[parentOne], rankTwo = this.rank[parentTwo];
        if(rankOne < rankTwo) {
            this.ultimateParent[parentOne] = parentTwo;
        }
        else if(rankOne > rankTwo) {
            this.ultimateParent[parentTwo] = parentOne;
        }
        else {
            this.ultimateParent[parentTwo] = parentOne;
            this.rank[parentOne] = rankOne + 1;
        }
    }

    public static void main(String[] args) {
        DisjointSet ds = new DisjointSet(7);
        int[][] edges = {{1, 2}, {2, 3}, {4, 5}, {6, 7}, {5, 6}, {3, 7}};

        for(int index = 0; index < edges.length - 1; index++) {
            ds.unionByRank(edges[index][0], edges[index][1]);
        }
        
        // Testing
        if(ds.findParent(3) == ds.findParent(7)) {
            System.out.println("Both are in same component");
        }
        else {
            System.out.println("Both are not in same component");
        }

        ds.unionByRank(3, 7);

        if(ds.findParent(3) == ds.findParent(7)) {
            System.out.println("Both are in same component");
        }
        else {
            System.out.println("Both are not in same component");
        }

    }
}

File: Disjoint Set/Python/ImplementationRank.py
# Time: O(4 alpha)
class disjointSet:
    
    def __init__(self, num):
        self.rank = [0] * (num + 1)
        self.ultimateParent = [index for index in range(num + 1)]
        
    
    def findParent(self, node):
        if node == self.ultimateParent[node]:
            return node
        self.ultimateParent[node] = self.findParent(self.ultimateParent[node])
        return self.ultimateParent[node]
    
    
    def unionByRank(self, nodeOne, nodeTwo):
        parentOne, parentTwo = self.findParent(nodeOne), self.findParent(nodeTwo)
        
        if parentOne == parentTwo:
            return
            
        rankOne, rankTwo = self.rank[parentOne], self.rank[parentTwo]
        
        if rankOne < rankTwo:
            self.ultimateParent[parentOne] = parentTwo
        elif rankOne > rankTwo:
            self.ultimateParent[parentTwo] = parentOne
        else:
            self.ultimateParent[parentTwo] = parentOne
            self.rank[parentOne] += 1
            

ds = disjointSet(7)
edges = [[1, 2], [2, 3], [4, 5], [6, 7], [5, 6]]

for one, two in edges:
    ds.unionByRank(one, two)
    
if ds.findParent(3) == ds.findParent(7):
    print("Both are in same component")
else:
    print("Both are not in same component")

ds.unionByRank(3, 7)

if ds.findParent(3) == ds.findParent(7):
    print("Both are in same component")
else:
    print("Both are not in same component")
